c/meterpreter/source/dump_sam/dump_sam.c

```diff
@@ -0,0 +1,304 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "dump_sam.h"
+#include "ReflectiveFreeAndExitThread.h"
+
+#define RDIDLL_NOEXPORT
+#define REFLECTIVEDLLINJECTION_CUSTOM_DLLMAIN
+#define REFLECTIVEDLLINJECTION_VIA_LOADREMOTELIBRARYR
+#include "ReflectiveLoader.c"
+
+
+/*! @brief Sets `dwResult` to the return value of `GetLastError()`, prints debug output, then does `break;` */
+#define BREAK_ON_ERROR( str ) { dwResult = GetLastError(); dprintf( "%s. error=%d (0x%x)", str, dwResult, (ULONG_PTR)dwResult ); break; }
+/*! @brief Sets `dwResult` to `error`, prints debug output, then `break;` */
+#define BREAK_WITH_ERROR( str, err ) { dwResult = err; dprintf( "%s. error=%d (0x%x)", str, dwResult, (ULONG_PTR)dwResult ); break; }
+
+/* Logging will work but only to OutputDebugStringA and not the full on Meterpreter logging because we don't have
+ * access to the API from within lsass.exe (which is where we're running).
+ */
+#ifdef DEBUGTRACE
+#define dprintf(...) real_dprintf(__VA_ARGS__)
+#if DEBUGTRACE == 1
+#define vdprintf dprintf
+#else
+#define vdprintf(...) do{}while(0);
+#endif
+#else
+#define dprintf(...) do{}while(0);
+#define vdprintf(...) do{}while(0);
+#endif
+
+/*!
+ * @brief Output a debug string to the debug console.
+ * @details The function emits debug strings via `OutputDebugStringA`, hence all messages can be viewed
+ *          using Visual Studio's _Output_ window, _DebugView_ from _SysInternals_, or _Windbg_.
+ */
+static _inline void real_dprintf(char* format, ...)
+{
+	va_list args;
+	char buffer[1024];
+	size_t len;
+	_snprintf_s(buffer, sizeof(buffer), sizeof(buffer) - 1, "[%04x] ", GetCurrentThreadId());
+	len = strlen(buffer);
+	va_start(args, format);
+	vsnprintf_s(buffer + len, sizeof(buffer) - len, sizeof(buffer) - len - 3, format, args);
+	strcat_s(buffer, sizeof(buffer), "\r\n");
+	OutputDebugStringA(buffer);
+	va_end(args);
+}
```

Comment:
You mentioned it already during module hacking, but I forgot the reason - what was the context behind not being able to write to the file based log file? 👀 

Was it just losing track of the original Metasploit config object that stores the config path to write to? Or are there quirks with writing to files on disk from lsass

Reply:
This code is what Meterpreter injects into LSASS so don't have access to the Meterpreter API from this context because we're in LSASS and not the Meterpreter process.

Reply:
Thanks; This is potentially something we'll want to wire up in the future. I'm not sure how windows handles two writers to the same file, from potentially different architectures. Probably best to leave it out of MVP either way :+1: